src/services/db.ts

```diff
@@ -118,6 +224,125 @@ export class WefwefDB extends Dexie {
       .limit(limit)
       .toArray();
   }
+
+  /*
+   * Settings
+   */
+
+  async populateDefaultSettings(tx: Transaction) {
+    const settingsTable = tx.table("settings");
+    settingsTable.bulkAdd(defaultSettings);
+  }
+
+  async migrateFromLocalStorageSettings(tx: Transaction) {
+    const localStorageMigrationKeys = {
+      font_size_multiplier: "appearance--font-size-multiplier",
+      use_system_font_size: "appearance--font-use-system",
+      collapse_comment_threads: "appearance--collapse-comment-threads",
+      post_appearance_type: "appearance--post-type",
+      use_system_dark_mode: "appearance--dark-use-system",
+      user_dark_mode: "appearance--dark-user-mode",
+    };
+
+    const settingsTable = tx.table("settings");
+
+    return await Promise.all(
+      Object.entries(localStorageMigrationKeys).map(
+        ([key, localStorageKey]) => {
+          const localStorageValue = localStorage.getItem(localStorageKey);
+
+          if (!localStorageValue) {
+            return Promise.resolve();
+          }
+
+          return settingsTable
+            .where(CompoundKeys.settings.key_and_user_handle_and_community)
+            .equals([key, "", ""])
+            .modify({
+              value: JSON.parse(localStorageValue),
+            });
+        }
+      )
+    );
+  }
+
+  getSetting<T extends keyof SettingValueTypes>(
+    key: T,
+    specificity?: {
+      user_handle?: string;
+      community?: string;
+    }
+  ) {
+    const { user_handle = "", community = "" } = specificity || {};
+
+    return this.transaction("r", this.settings, async () => {
+      let setting = await this.settings
+        // Everything matches
+        .where(CompoundKeys.settings.key_and_user_handle_and_community)
+        .equals([key, user_handle, community])
+        .first();
+
+      if (!setting && user_handle !== "") {
```

Comment:
I believe this condition should be community, not user_handle (and vice versa below). That way you don't clear a string that is already empty, because that's a noop and it will be a waste of a query.

Reply:
I don't think I get this comment. I'm intending to do:
- If setting isn't found AND user handle is not empty,
- Try to get user setting for this key
And the same thing for the community below. Am I missing something here?

Reply:
My point was that if community handle is empty, you already ran that query above. See the diff for ebac2cf87a.

Reply:
Except I but the if statement on 289 around the wrong thing, it should be on 297.

Reply:
Basically I think you should only run the inner two queries if both the user handle and community are set, otherwise the original query and the final one with both empty will cover everything.

Reply:
But when doing the first query we don't check if the community and user are empty or not. They may be non-empty when we arrive at this step. Then inside this condition we try as if community is empty to see if there is a user setting for  this key. 
I checked the diff from your branch but I still think this is correct logic. Consider following scenarios:
``` typescript
getSetting("show_nsfw", { user: "test_user", community: "test_community" }); // 1
getSetting("show_nsfw", { community: "test_community" }); // 2
getSetting("show_nsfw", { user: "test_user" }); // 3
getSetting("show_nsfw") // 4
```
1st call:
- Tries with key+handle+community, if there is a match returns it
- If there is no match AND handle is NOT empty; tries with `key+handle+[empty_community]`. This is a user setting without a specific community.
- If there is no match AND community is NOT empty; tries with `key+[empty handle]+community`. This is a community setting without a specific handle.
- If there is still no match, then tries with only the key.

2nd call:
 - Tries with key+handle+community, if there is a match returns it
 - no setting found and handle is empty, skips this step.
 - no setting found and community is NOT empty, tries with `key+[empty handle]+community`.
 - If there is still no match, then tries with only the key.

3rd call:
 - Tries with key+handle+community, if there is a match returns it
- If there is no match AND handle is NOT empty; tries with `key+handle+[empty_community]`. This is a user setting without a specific community.
 - no setting found and community is empty, skips this step.
 - If there is still no match, then tries with only the key.

4rd call:
- The first query will succeed `key+[empty handle]+[empty community]`

Reply:
Yeah in that case we do an extra unnecessary query. We can eliminate that with an additional empty check for community (or user handle in the next one). But apart from the extra query, the logic should be correct. 
Let me update that part.

Reply:
yeah @rsammelson you're right. Noticed it while updating the code. Check the new version please.